### ISSUE TITLE
Tracking Issue for trim-paths RFC 3127

### ISSUE THREAD

**Issue by ehuss** [rust-lang/rust#111540]
This is a tracking issue for the RFC 3127 (rust-lang/rfcs#3127).

Issues: https://github.com/rust-lang/rust/labels/F-trim-paths
Implementation: TODO
Documentation: TODO

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Implement the RFC (cc @rust-lang/t-compiler -- can anyone write up mentoring instructions?)
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

- [ ] Should we use a slightly more complex remapping rule, like distinguishing packages from registry, git and path, as proposed in https://github.com/rust-lang/rust/issues/40552?
- [ ] What concerns are there regarding the use of relative paths causing issues with backtraces and debuggers?
- [ ] Should we treat the current package the same as other packages?
- [ ] Will these cover all potentially embedded paths? Have we missed anything?

**Comment by weihanglo** [comment]
I was trying to remove `OSO` symbols on macOS. Did a research and summarized in this issue:  <https://github.com/rust-lang/rust/issues/116948#issuecomment-1793617018>. Now, I can see that there is no way for `trim-paths` to really trim all paths unless rustc does some dirty trick for older macOS platforms. Would it be a blocker if we want to stabilize this feature, say, in 2024 edition?

cc @Urgau you might be interested.

**Comment by michaelwoerister** [comment]
> What concerns are there regarding the use of relative paths causing issues with backtraces and debuggers?

I talked about this with @danielframpton the other day and he noticed that the current `[package name]-[version]` rule for remapping might make things unnecessarily complicated for debuggers. For making the debugging experience easy to set up, it would be beneficial to require as little configuration as possible. The RFC does not explicitly make file system paths (as used by Cargo) match with how things are renamed, so in the general case the debugger needs a separate source file search path for each crate involved. 

However, if we prescribed that each crate from was remapped to the relative path Cargo uses inside its `$CARGO_HOME/registry/src` and `$CARGO_HOME/git/checkouts` directories, then we have a fixed number of source directories a debugger needs to know about. The renaming would then be `[registry]/[package name]-[version]` for registry crates and `[package name]-[sha]` for git dependencies.

The current `[package name]-[version]` rule is already pretty close, but that seems to be by accident. It would be good to make this something that can be relied on. 

This might also help with backtraces?

I don't know if there is something similarly useful we can do for path dependencies.

> Should we treat the current package the same as other packages?

This also came up. Unconditionally stripping the path to the current package might be too inflexible in some cases. E.g. when building a staticlib, one might want to make the paths coming from the current package identifiable in some way. If they all start with `src/..` then they might clash with the paths of the codebase that uses the staticlib as a dependency.

**Comment by weihanglo** [comment]
> However, if we prescribed that each crate from was remapped to the relative path Cargo uses inside its `$CARGO_HOME/registry/src` and `$CARGO_HOME/git/checkouts` directories, then we have a fixed number of source directories a debugger needs to know about. The renaming would then be `[registry]/[package name]-[version]` for registry crates and `[package name]-[sha]` for git dependencies.

This plan sounds really tenable to me! Will do a change on cargo side for this, and see how people think about it.

For path dependencies, maybe we could have a common prefix for them, like `cargo_deps@local/<pkg>-<version>`. Might not be useful as others dep kinds, but at least people can have a single location for path dependencies?

> then they might clash with the paths of the codebase that uses the staticlib as a dependency.

Sorry I don't fully understand. Could you give a more concrete example of this?

**Comment by michaelwoerister** [comment]
> For path dependencies, maybe we could have a common prefix for them, like cargo_deps@local/<pkg>-<version>. Might not be useful as others dep kinds, but at least people can have a single location for path dependencies?

Maybe, yes. It's definitely worth thinking about this some more (and collecting ideas from other people).

> Sorry I don't fully understand. Could you give a more concrete example of this?

As I understand it, the RFC states that paths from the workspace being compiled are trimmed by simply making them relative to the workspace root, that is, all file paths from workspace crates would look like `src/foo/bar.rs` or `some_crate_in_the_workspace/src/foo/bar.rs`. Those paths don't contain anything that would allow disambiguating them from other paths that are purely relative. One example where this is sub-optimal is when one has multiple Rust projects, each of which gets compiled into a staticlib that is then consumed within another, non-Rust project. Each of these Rust projects would have debuginfo file paths that just start with `src/` and the paths for some of the files might clash , e.g. `src/lib.rs`. The debugger would have no way to know which `src/lib.rs` file to pick, even if both are in its source search path.

Therefore it would be good, if there was some way to avoid this problem. Allowing to prepend a prefix to the relative paths might be an option.

**Comment by weihanglo** [comment]
@michaelwoerister. Thank you for your feedback! I just opened a new issue <https://github.com/rust-lang/cargo/issues/13171> for further discussion about remap rules in Cargo.

**Comment by michaelwoerister** [comment]
Hi everyone, while discussing [@weihanglo's PR](https://github.com/rust-lang/rust/pull/118518) that tries to fix some debuginfo related issues, [it started to look like](https://github.com/rust-lang/rust/pull/118518#discussion_r1464656443) there might be some limits to how accurate the debuginfo related scopes can be implemented. It seems like we'll have a hard time controlling what ends up in split- vs in unsplit-debuginfo. 

I'm wondering: do we really need three debuginfo-related scopes? Or can we just have a single debuginfo scope? Do we have a concrete use case that requires the fine granularity of three scopes?

(cc @cbeuw)

**Comment by cbeuw** [comment]
There were requests from people who want to debug crash dumps in shipped binaries or profile them: https://github.com/rust-lang/rfcs/pull/3127#issuecomment-850387079 and https://github.com/rust-lang/rfcs/pull/3127#issuecomment-850491441.

Before settling on two debuginfo scopes, the original design to address this had a single debuginfo scope in rustc and made it Cargo's job to emit it (under the release profile) when debuginfo splitting is off, and to omit it when debuginfo splitting is on. The rationale behind having separate scopes (https://github.com/rust-lang/rfcs/pull/3127#discussion_r857881753) I believe was to reduce the need for special casing on Cargo's side.

Seeing that the complexity we gain in rustc is overwhelming the complexity we would've saved in Cargo, I think it's fine to go back to the previous approach which is making debuginfo remapping all-or-nothing.

However, under this design, rustc may be invoked with debuginfo splitting and no debuginfo path remapping. Therefore LLVM _must not_ emit debuginfo paths anywhere in distributable files, otherwise we'd fail privacy and reproducibility. Is this possible? With separate scopes, at least rustc can know that unsplit debuginfo needs to be remapped and can babysit LLVM somewhat. rustc won't know this with a single debuginfo scope.

**Comment by michaelwoerister** [comment]
Thanks for the pointers, @cbeuw! I can see the rationale behind all of these concerns. It might be a tricky constellation of tradeoffs. 

Taking a step back, I'm not sure if the scopes as they are defined now are that useful in practice. Right now, they are defined in terms of the *kind of data* that belongs to them (e.g. diagnostics, debuginfo, file!() macro expansions). But in practice, I think, it would make sense to define them in terms of which output artifacts they refer to. In other words, there would be three main scopes:
- console output (~= diagnostics)
- the binary (~= macro expansions and debuginfo)
- separate debuginfo

That would make it much easier for a user to decide what they need.

I'm also not quite clear on whether it is possible to have a completely sanitized binary together with an unsanitized separate debuginfo file. That is, does that even solve the usability problem? Can tools find the debuginfo if they get a sanitized binary? I think that question needs to be answered per tool and per platform, because the answer could be different for each combination. It would be good to have concrete use cases that we can test our implementation against.

**Comment by michaelwoerister** [comment]
I did some reading and [some testing](https://github.com/michaelwoerister/split-debuginfo-sanitization) and it looks to me like split-dwarf cannot be used to sanitize binaries while keeping an unsanitized, separate debuginfo file for a good out-of-the-box debugging experience. The binary still contains the `.debug_line` section which contains the paths to all the source files (see e.g. `7.3.2.1 First Partition (with Skeleton Unit)` in the [DWARF 5 standard](https://dwarfstd.org/doc/DWARF5.pdf)). So, if we want a sanitized binary, we have to sanitize these paths. But that breaks the out-of-the-box debugging experience. At that point the distinction between the split-debuginfo and unsplit-debuginfo scopes does not make much practical difference anymore, I think.

There is another way for separating out debuginfo via postprocessing:

```bash
objcopy --only-keep-debug my_binary my_binary.debug
strip -S my_binary -o my_binary.stripped
objcopy --add-gnu-debuglink=my_binary.debug my_binary.stripped
```

That approach really can sanitize the binary while keeping debuginfo intact. But we would have to make that the default for release builds in order to preserve the out-of-the-box debugging experience. I'm not sure how viable that is.

**Comment by michaelwoerister** [comment]
I think it is a good idea to revisit the question of whether we really want the default for `--release` builds to be that the binary is sanitized but separate debuginfo files aren't. It seems to me that the whole point of sanitizing by default is to avoid unpleasant surprises when shipping `--release` artifacts. Yet, only sanitizing half of the artifacts seems to undermine that goal of not causing any surprises. My suggestion would be to either sanitize everything or nothing by default.

An additional reason not go the mixed-sanitization route is that it does not actually solve the problem on Linux because [with split-dwarf the relevant parts stay in the binary](https://github.com/rust-lang/rust/issues/111540#issuecomment-1923586433) (at least as far as I can tell). 

Regarding the debugging experience: There is [an ongoing discussion](https://github.com/rust-lang/cargo/issues/13171) one how to make that as seamless as possible even when paths are trimmed. We haven't landed on a perfect solution yet. 

Possible solutions I see are:
 - Use a dev-opt Cargo profile that can be used for locally debugging and profiling.
 - Somehow promote the [bench profile](https://doc.rust-lang.org/cargo/reference/profiles.html#bench) for that use-case.
 - Have a `cargo debug` subcommand that configures the debugger's source paths.

cc'ing some folks who commented on not trimming separate debuginfo files in the RFC thread: @joshtriplett, @BurntSushi, @kornelski

**Comment by kornelski** [comment]
For me this is very profile-dependent, so I'd be happy if Cargo had good profile-dependent defaults.

* in `dev` and `bench` profiles I need the debug info working reliably and being maximally compatible. I never ship these, so I'm fine with full paths, even if they include my home dir and `~/.cargo`'s guts. I'd prefer debug info to just work in debuggers and profilers without configuring any search directories or mappings.

* In release profile, I don't want full paths. I prefer smaller binaries and more privacy/obscurity. In release binaries I only care about getting stack traces, and for that function name + file name is enough. I like having an option to split debuginfo out to external files, so that I can make it an optional install.

**Comment by bjorn3** [comment]
For `cargo install` I wouldn't want trim paths. And ditto for precompiled opensource binaries as created using eg cargo-dist and downloaded using cargo-binstall.

**Comment by bjorn3** [comment]
Every once in a while I have a binary that panics or hangs. I want to be able to find out why without having to recompile it from scratch.

**Comment by ssokolow** [comment]
Whereas, in my case, I'm operating on the same "Why pessimize the 99.9% to save time on the 0.1% situations?" logic as the suggestion to use `debug=0` and `strip = "debuginfo"` to speed up link times at https://davidlattimore.github.io/working-on-rust-iteration-time.html

**Comment by ssokolow** [comment]
...plus, as a developer, I'd *prefer* that users be required to re-`cargo install` to go into a debugging configuration so that, if I've made a new release since whatever they're running, they'll have an opportunity to encounter any potentially relevant bugfixes I've already done.

**Comment by michaelwoerister** [comment]
> For cargo install I wouldn't want trim paths. And ditto for precompiled opensource binaries as created using eg cargo-dist and downloaded using cargo-binstall.

To clarify: trim-paths will still give file paths of the form `some_crate-1.0.1/src/lib.rs`, so the information in a backtrace is still rather useful. If debuginfo is enabled (which currently is not the default for release builds), then the debugger should still be able to find the symbols, it just won't be able to find the source files currently.

**Comment by michaelwoerister** [comment]
Here is a table that shows the various combinations of the `trim-paths` and `split-debuginfo` Cargo options and (to the right side of the `⇒`) what `rustc` options these map to and what the consequences of this are.
 
trim-paths | split-debuginfo | ⇒ | --remap-path-scope | Binary | Separate Debuginfo | Surprising? | Useful? | Remarks
-- | -- | -- | -- | -- | -- | -- | -- | --
none | off | ⇒ |   | unsanitzed |   |   | Yes |  
macro | off | ⇒ | macro | semi-sanitized |   |   | ? |  
object | off | ⇒ | macro + split-debuginfo + split-debuginfo-path + unsplit-debuginfo | sanitized |   | Kind of | ? | Must sanitize debuginfo too, otherwise "object" won't be sanitized
all | off | ⇒ | macro + split-debuginfo + split-debuginfo-path + unsplit-debuginfo | sanitized |   |   | Yes |  
 |  |  |  |  |  |  |  | 
**trim-paths** | **split-debuginfo** | ⇒ | **--remap-path-scope** | **Binary** | **Separate Debuginfo** | **Surprising?** | **Useful?** | **Remarks**
none | packed | ⇒ |   | unsanitzed | unsanitzed |   | Yes |  
macro | packed | ⇒ | macro | semi-sanitized | unsanitzed |   | ? |  
object | packed | ⇒ | macro + unsplit-debuginfo + split-debuginfo-path | sanitized | unsanitzed | Yes | ? | Debugger won't find source files
all | packed | ⇒ | all | sanitized | sanitized |   | Yes (?) | Debugger won't find source files
|  |  |  |  |  |  |  | 
**trim-paths** | **split-debuginfo** | ⇒ | **--remap-path-scope** | **Binary** | **Separate Debuginfo** | **Surprising?** | **Useful?** | **Remarks**
none | unpacked | ⇒ |   | unsanitzed | unsanitzed |   | Yes | Fast builds with debuginfo
macro | unpacked | ⇒ | macro | semi-sanitized | unsanitzed |   | ? |  
object | unpacked | ⇒ | macro + unsplit-debuginfo + split-debuginfo-path | sanitized | unsanitzed | Yes | ? | Debugger won't find source files AND dwos
all | unpacked | ⇒ | all | sanitized | sanitized |   | ? | Debugger won't find source files AND dwos
|  |  |  |  |  |  |  | 
**trim-paths** | **split-debuginfo** | ⇒ | **--remap-path-scope** | **Binary** | **Separate Debuginfo** | **Surprising?** | **Useful?** | **Remarks**
none | [post-link](https://github.com/rust-lang/compiler-team/issues/721) | ⇒ |   | unsanitzed | unsanitzed |   | Kind of | Not much use in splitting
macro | post-link | ⇒ | macro | sanitized | unsanitzed | A little | Kind of |  
object | post-link | ⇒ | macro + unsplit-debuginfo + split-debuginfo-path + split-debuginfo | sanitized | unsanitzed |   | Kind of | Might be useful when only distributing binary?
all | post-link | ⇒ | all | sanitized | sanitized |   | Yes |  

Observations:
- Some of the combinations don't have a real use case
- Some of the combinations are surprising. They don't do what one this they would do. E.g. `trim-paths=object, split-debuginfo=packed` leads to semi-broken debuginfo because line-tables, contained in the binary, have to be sanitized. But it the combination sounds like "don't touch debuginfo".
- Only two cases make a distinction between the `split-debuginfo` and `unsplit-debuginfo` remapping scopes -- and they fall into the surprising and useless category.
- No case makes a distinction between `unsplit-debuginfo` and `split-debuginfo-path`

My conclusion:
- We should only have a single `debuginfo` scope and not further subdivide that into `split-debuginfo`, `unsplit-debuginfo`, and `split-debuginfo-path`. The only cases this makes a difference for are `trim-paths=object, split-debuginfo=packed` and `trim-paths=object, split-debuginfo=unpacked` but these cases have no real use case to begin with.

**Comment by Zalathar** [comment]
Via #122450, it’s come to my attention that this will affect coverage instrumentation, because the coverage mappings embedded in the binary usually contain at least one absolute path, and any adjustment to those paths is constrained by the capabilities of the `llvm-cov` tool.

**Comment by michaelwoerister** [comment]
For reference, Clang allows to separately control path remapping for code coverage: https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fcoverage-prefix-map

**Comment by redactedontop** [comment]
...updates?

**Comment by liciana24** [comment]
any movement here? still waiting on this and hoping scope didnt change from stripping all in release mode BY DEFAULT.

**Comment by weihanglo** [comment]
Not entirely sure on rustc side of this, but I feel like it is waiting for Cargo to report back the integration status. See also: <https://github.com/rust-lang/cargo/issues/12137#issuecomment-2149647027>

**Comment by estebank** [comment]
Random question: are these flags meant to affect the output of diagnostics as well? One could argue that they should.

**Comment by Urgau** [comment]
Yes, that's the `diagnostics` scope, see the [`--remap-path-scope`](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/remap-path-scope.html) flag for more details.

**Comment by lina9992006** [comment]
status?

**Comment by weihanglo** [comment]
It is interesting that `include!(concat!(env!("OUTDIR"), "…"))` combo got leaked, though I can't reproduce it on Linux with the small reproducer below, even for a debug build. @teknalb Does it only happen on Windows?

<details><summary>Details</summary>
<p>

```rust
// build.rs
fn main() {
    let dest = std::path::PathBuf::from(std::env::var("OUT_DIR").unwrap());
    std::fs::write(
        dest.join("bindings.rs"),
        r#"pub fn yes(n: u64) { println!("yes {n}!!", ); }"#,
    )
    .unwrap();
}
```

```rust
// src/lib.rs
include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
```

```rust
// src/main.rs
fn main() {
    bar::yes(2);
}
```

</p>
</details>

**Comment by weihanglo** [comment]
@teknalb It might be worth opening a separate issue report with a full reproducible example, including how you figured out those embedded strings. `profile.trim-paths = true` is inherently the stable `--remap-path-prefix` flag without any scope, so it might be a regression or bug. I had a hard time reproducing with the minimal reproducer here <https://github.com/rust-lang/rust/issues/111540#issuecomment-2524437598> on Windows with `strings` from Sysinternals. Mind trying out the minimal reproducer and reporting back the result? Maybe we it is not a good one.

There are indeed absolute paths in pdb file itself, but not in the final executables. Despite that it sounds like a separate issue.

I feel like it's not possible to remove absolute paths in pdb, which `--remap-path-scope=debuginfo` ought to support. However, I saw [the `PathMap` flag](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/advanced#pathmap) in C# compiler, so we still have hope. @ChrisDenton Do you have any suggestion?

**Comment by tabbyrobin** [comment]
Will `--remap-path-scope` address the presence of completely random paths (such as temp directories)?

For example, the ones that `uv build` generates: https://github.com/astral-sh/uv/issues/13096

**Comment by Urgau** [comment]
No, this would be done with `--remap-path-prefix`.

`--remap-path-scope` only permits changing the scope of the remapping (diagnostic & macro, debuginfo only, ...).

**Comment by teknalb** [comment]
I would like to ask if this feature concerns domain names and IP addresses as well.

![Image](https://github.com/user-attachments/assets/a4a3e71b-efd8-4e80-b5ee-2a590535fd9f)

I run a `gitea` cargo registry on a local network and the server's IP is all over the place, might not be important because it's on local network but that could be concerning for remote private servers.

**Comment by weihanglo** [comment]
@teknalb that part is controlled by Cargo. Let's discuss in <https://github.com/rust-lang/cargo/issues/13171> then!

**Comment by csmulhern** [comment]
I want to raise that I think it should be possible to apply different remapping to the same path for different scopes. It seems like remap-path-prefix was too generic in remapping across many different use cases, and now the scope argument is being introduced to limit the scope, but that doesn't deal with the situation where you want to remap the same path differently across scopes.

E.g. imagine I want to remap `/home/cameron/code/project` to `"."` for `macro`, but to `""` for `debuginfo`. This would be necessary on macOS, where OSO stabs cannot use relative paths (see #132143).

You could imagine doing something like:

```bash
rustc ...
  --remap-path-scope=debuginfo \
  --remap-path-prefix=/home/cameron/code/project=. \
  --remap-path-scope=macro \
  --remap-path-prefix=/home/cameron/code/project=
```

But repeated usage of these flags are independently aggregated, rather than ordered, so that wouldn't work.

Not sure on the right design here, but I do think being able to express ^ is worthwhile.

**Comment by weihanglo** [comment]
@ulrichard 
That would likely be an issue of how Cargo determine rustc metadata hash. This tracking issue is for rustc to trim paths in binaries only. Please open issue in <https://github.com/rust-lang/cargo/issues/new/choose> with steps to repro or at least support materials that maintainers can look at. Thank you.

**Comment by weihanglo** [comment]
FWIW, Cargo part of this feature is calling for testing. For more details, see <https://github.com/rust-lang/cargo/issues/12137#issuecomment-5607218160>.
